Mecanum wheels demo

scpeters opened the pull request:
# 🎉 New feature

Closes https://github.com/ignitionrobotics/ign-gazebo/issues/579

## Summary

This adds an example world that is based on the diff_drive_skid example but uses a new feature in dartsim exposed by an SDFormat attribute in a custom xml namespace in https://github.com/ignitionrobotics/ign-physics/pull/224. The friction parameters are modified in 48010ca to use a friction direction that is constant in the chassis frame and points at a 45 degree angle from each wheel's revolute joint axis. The friction coefficients are set to 1.0 and 0.0 in the perpendicular directions.

## Test it

The vehicle can be operated like a skid-steer vehicle using the diff-drive system that is currently attached. I need to create a new system to enable driving sideways.

To see it move sideways, you can do the following:

1. comment out the diff-drive plugin
2. modify the gravity vector to have a non-zero value in the x or y direction
3. rotate the vehicle so that the x/y gravity component is pointing in the lateral direction of the vehicle
4. watch it go sideways

## Checklist
- [X] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

chapulina commented on examples/worlds/diff_drive_mecanum.sdf:
```diff
@@ -0,0 +1,362 @@
+            <!--sphere>
+              <radius>0.3</radius>
+            </sphere-->
+            <box><size>0.6 0.6 0.6</size></box>
```
:sweat_smile: 

![mecanum](https://user-images.githubusercontent.com/5751272/112069340-e6ad4580-8b28-11eb-8fcb-973f73b7b181.gif)

scpeters replied:
:sweat_smile: *nervous laughter*

scpeters replied:
reverted to spheres in 53792df

chapulina replied:
The box was useful to see the directions in which the wheels were rotating

chapulina commented on src/systems/mecanum_drive/MecanumDrive.hh:
```diff
@@ -0,0 +1,116 @@
+  // Forward declaration
+  class MecanumDrivePrivate;
+
+  /// \brief Differential drive controller which can be attached to a model
```
Reminder to update docs

scpeters replied:
095e0d5

chapulina commented on src/systems/mecanum_drive/SpeedLimiter.cc:
```diff
@@ -0,0 +1,200 @@
+/*********************************************************************
```
We could use the new `ignition::math::SpeedLimiter` here, or just leave this one in for now and update later.

scpeters replied:
good call

scpeters replied:
9bbcceb

scpeters commented:
ok, I just got a working joystick configuration with an ign-launch file in https://github.com/scpeters/ign-launch/commit/ad0efb2b4f0907d993ebd40788998d4798064893

on my gamepad, the left analog stick commands longitudinal and lateral velocity, and the horizontal part of the right analog stick is used for turning

scpeters commented:
I think this is ready for review. It's missing the Odometry, but the controller works

azeey commented on examples/worlds/diff_drive_mecanum.sdf:
```diff
@@ -0,0 +1,362 @@
+<?xml version="1.0" ?>
+<!--
+  Ignition Gazebo skid steer differential drive plugin demo
```
nit: Can you update these comments?

scpeters replied:
15416d0

azeey commented on src/systems/mecanum_drive/MecanumDrive.cc:
```diff
@@ -0,0 +1,719 @@
+  }
+
+  // If the joints haven't been identified yet, look for them
+  static std::set<std::string> warnedModels;
```
Not sure why `warnedModels` is a set. Can there be more than one model here?

azeey commented on src/systems/mecanum_drive/MecanumDrive.cc:
```diff
@@ -0,0 +1,719 @@
+
+  // constant used in computing target velocities
+  const double angularLength = 0.5 * (this->wheelSeparation + this->wheelbase);
+  const double invWheelRadius = 1 / this->wheelRadius;
```
Can `wheelRadius` be 0?

azeey reviewed (approved):
This is so cool!! Left a few minor comments. If you don't have time to implement the odometry, I'd suggest removing the functionality completely and revisiting later.

chapulina commented on src/systems/mecanum_drive/MecanumDrive.cc:
```diff
@@ -0,0 +1,719 @@
+  {
+    double minVel = _sdf->Get<double>("min_velocity");
+    this->dataPtr->limiterLin->SetMinVelocity(minVel);
+    this->dataPtr->limiterAng->SetMinVelocity(minVel);
```
I realize that this is just following the pattern from other systems, but I'm wondering if we should split linear and angular components, considering that they have different units.

chapulina commented on src/systems/mecanum_drive/MecanumDrive.cc:
```diff
@@ -0,0 +1,719 @@
+
+  double linVel;
+  double latVel;
+  double angVel;
```
Small suggestion, instead of 3 vars, this could use a `Commands` object

chapulina commented on examples/worlds/mecanum_drive.sdf:
```diff
@@ -0,0 +1,365 @@
+
+  Try sending commands:
+
+    ign topic -t "/model/vehicle_blue/cmd_vel" -m ignition.msgs.Twist -p "linear: {x: 0.5}, angular: {z: 0.05}"
```
It would be interesting to use a command that has `y` linear velocity.

scpeters replied:
added in 15416d0

chapulina reviewed (commented):
It works! Just some smaller comments from @azeey and me.

![mecanum_lat](https://user-images.githubusercontent.com/5751272/112266539-6292b500-8c31-11eb-9c60-2332b976ccf0.gif)

scpeters commented:
> If you don't have time to implement the odometry, I'd suggest removing the functionality completely and revisiting later.

removed in 49eb8cb

scpeters commented on examples/worlds/diff_drive_mecanum.sdf:
```diff
@@ -0,0 +1,362 @@
+      </link>
+    </model>
+
+    <model name='vehicle_blue' xmlns:gazebo="http://gazebosim.org/schema">
```
fixed in 15416d0

scpeters commented on examples/worlds/mecanum_drive.sdf:
```diff
@@ -0,0 +1,365 @@
+<?xml version="1.0" ?>
```
I renamed the other file in bcfb881 to `skid_steer_mecanum`. I think it's still interesting, since it uses the diff drive plugin for all the wheels just like the `diff_drive_skid` example world and behaves like a skid steer, but I think it has a smoother response. I think it's interesting, but I can delete it if it's too much noise

scpeters commented:
I haven't yet addressed some of the comments related to the code in MecanumDrive. I wasn't very thoughtful in architecting it, I just copied DiffDrive and added to it. I'm not sure I'll have time to address those concerns before code freeze given other higher priority pull requests. Maybe addressing the current set of comments is doable, but I'm sure there are more and more issues to be found here if reviews continue getting stricter and I'd rather focus elsewhere than polishing this one

chapulina reviewed (approved):
LGTM as a first revision! I'll come back to this after the release and address the remaining comments, which should be backwards compatible.
